src/main/java/org/truffleruby/core/thread/ThreadManager.java

```diff
@@ -232,10 +234,35 @@ public void initialize(DynamicObject rubyThread, Node currentNode, String info,
         Layouts.THREAD.setSourceLocation(rubyThread, info);
 
         final Thread thread = createJavaThread(() -> threadMain(rubyThread, currentNode, task));
+
         thread.setName(NAME_PREFIX + " id=" + thread.getId() + " from " + info);
+
+        final CompletableFuture<Throwable> thrown = new CompletableFuture<>();
+
+        thread.setUncaughtExceptionHandler((t, e) -> {
+            thrown.complete(e);
+        });
+
         thread.start();
 
-        FiberManager.waitForInitialization(context, Layouts.THREAD.getFiberManager(rubyThread).getRootFiber(), currentNode);
+        if (!FiberManager.waitForInitialization(context, Layouts.THREAD.getFiberManager(rubyThread).getRootFiber(), currentNode)) {
+            final String message;
+
+            try {
+                if (thrown.isDone()
+                        && thrown.get() instanceof IllegalStateException
+                        && thrown.get().getMessage().startsWith("Multi threaded access requested")
+                        && context.getEnv().isCreateThreadAllowed()) {
+                    message = thrown.get().getMessage() + " Are you attempting to create a Ruby thread after you have used a single-threaded language?";
+                } else {
+                    message = "creating thread timed out";
+                }
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
+            }
+
+            throw new RaiseException(context.getCoreExceptions().securityError(message, currentNode));
+        }
```

Comment:
We could deal with this much more nicely, and without depending on setting the UncaughtExceptionHandler by creating our own `Thread` instance and add a try/catch around the TruffleContext.enter()/leave().

Reply:
I thought `context.getEnv().createThread` was the proper way to create a thread?

Reply:
And the documentation actually *recommends* using `setUncaughtExceptionHandler`.

Reply:
Both work, the createThread() API was introduced first, but is much less flexible than doing our own thread+exception handling. The communication with `setUncaughtExceptionHandler()` looks quite ugly to me.